Add missing options, add method arguments lists with descriptions, examples and definitions

admusin opened the pull request:
Add code tags around names of: methods, arguments, attributes, options, HTML elements, where missing.
Add argument list with description for each method on the page
Add list of used templates after method description, in all method where it applies.
Add Form::create() options list with description in new subsection
Add missing options in options list for: input() and add details in some option definitions.
Reword Input Options descriptions to keep current definition but start by option type specification.
Clarify in introduction of Common Options for Specific Inputs that the options will be provided either in $options or in $attributes depending on the method
Create subsection for Checkbox, Select and Radio and move common Options for Select, Checkbox and Radio at beginning of that subsection for easier reference.
Clarify in Options for Select, Checkbox and Radio when 'value' and 'empty' will do for each method
Indent code block examples inside all options lists, one level further, in order to allow faster and easier scan of the list and differentiation between definitions/descriptions and examples.
Create subsection for Date & Time controls and move common Options for Date & Time at beginning of this subsection for easier reference.
Separate Options for Date & Time in 3 sub-subsections: Common Options for Date & Time, Options for Date and Options for Time
Add missing Date & Time options from API source code comments, along with explanations.
Add missing date() method subsection inside the Date & Time section along with example.
Improve explanations of Date & Time options based on comments in API source code.
Add ref links from each Date & Time method arguments description towards: common input options, common datetime options and respectively common date or time options (depending on method).
Transform textual note at the end of Creating Month Inputs into reST note.
Add example output for minutes interval example code, in Creating Minute Inputs.
Regroup options specific to any one particular method inside a list in the method's section; easier to identify and skim through.
Add 2 subsections inside Displaying and Checking errors; one for each method for easier reference from TOC.
Create 2 subsections inside Creating Buttons and Submit Elements: one for submits and one for buttons; allows easier access from TOC.
Create 2 subsections inside Creating Standalone Buttons and POST Links; allows easier access from TOC.
Create 2 subsections inside Generating Entire Forms: allows easier access from TOC.
Add reST note about additional options passed to Submit, being passed to the input element.
Elaborate definition in Creating Submit Elements subsection, based on API source-code comments.
Change last example code block inside inputs() method to use the actual method instead of allInputs().
Add reST note in Creating Form Inputs to remind that elements generated by input() are always wrapped in divs.
Elaborate description for error() and add reST tip about use of validator messages.
Improve and expand some definitions and explanations across page, based on detailed comments from API source code.
Expand explanation introducing the autodetected datatypes in Creating Form Inputs.
Add links to context provider classes in API from each entry in the list of built-in context providers in Starting a Form section.
Clarify in Moving Checkboxes and Radios Outside of a Label that only checkboxes created via input() are wrapped in a label.
Add details about radio button creation methods in Moving Checkboxes and Radios Outside of a Label

markstory commented on en/views/helpers/form.rst:
```diff
@@ -16,39 +16,53 @@ Starting a Form
+``$model``:
+
+* An ``Entity`` instance or an iterator will map to
+  `EntityContext <https://api.cakephp.org/3.3/class-Cake.View.Form.EntityContext.html>`;
```
You're missing `_` on the end of this link.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -16,39 +16,53 @@ Starting a Form
+  built-in ORM.
+
+* An array containing the ``'schema'`` key, will map to
+  `ArrayContext <https://api.cakephp.org/3.3/class-Cake.View.Form.ArrayContext.html>`
```
And this one.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -16,39 +16,53 @@ Starting a Form
+  which allows you to create simple data structures to build forms against.
+
+* ``null`` and ``false`` will map to
+  ``NullContext <https://api.cakephp.org/3.3/class-Cake.View.Form.NullContext.html>``;
```
This link is also incorrect.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -91,18 +107,62 @@ Output:
+    Since this is an *edit* form, a hidden ``input`` field is generated to
     override the default HTTP method.
 
+Options for Form
```
Should this mention its for form *creation*

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -91,18 +107,62 @@ Output:
 -----------------------------------
 
-By using the ``type`` option you can change the HTTP method a form will use::
+By using the ``'type'`` option you can change the HTTP method a form will use::
```
We don't generally include single quotes around key name as the inline code indicates that it is a key name.

admusin replied:
I've removed the quotes around this word as indicated. 
I think I've added quotes around pretty much every key name; if I understand correctly I should remove them everywhere, right?
As a more particular case, should I remove them even in the enumeration lists (i.e. where we show a list of all the available key names, with their description; I'm asking that since in those places I guess they are meant to indicate what we should literally input in our code and might be helpful to leave no room for guessing for beginners.

My rationale for adding them was that some trivial key names such "type", "empty", "default", "value" etc. coincide with standard vocabulary words as well as with certain HTML attributes and I thought it would help disambiguate what we refer to, since it certain phrase constructions it can become fuzzy to tell what the author's intended meaning is (it might be more obvious for a veteran Cake user, but then again, they're not the main target audience of the manual as they need it less ).
Of course, if you think they need to go away, everywhere, I'll remove them all over the page where code examples aren't involved. Just making sure what I should do, for now.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -233,13 +298,38 @@ Creating Form Inputs
-necessary. By using the metadata in the form context, this method will choose an
-appropriate input type for each field. Internally ``input()`` uses the other
-methods of FormHelper.
+* ``$fieldName`` - A field name in the form ``'modelname.fieldname'``.
```
`modelname` should probably be capitalized

admusin replied:
Thanks for pointing that out.
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -233,13 +298,38 @@ Creating Form Inputs
+By default the ``input()`` method will employ the following widget templates::
+
+'inputContainer' => '<div class="input {{type}}{{required}}">{{content}}</div>'
+'input' => '<input type="{{type}}" name="{{name}}"{{attrs}}/>'
```
This block should be indented.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -233,13 +298,38 @@ Creating Form Inputs
+
+In case of validation errors it will also use::
+
+'inputContainerError' => '<div class="input {{type}}{{required}} error">{{content}}{{error}}</div>'
```
This block should be indented as well.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -315,31 +414,35 @@ A more extensive example showing some options for a date field::
+If your model name consists of two or more words (e.g.
+"UserGroups"), when passing the data using ``set()`` you should name your
+data in a pluralised and
+`lower camelCased <https://en.wikipedia.org/wiki/Camel_case#Variations_and_synonyms>`
```
This link is missing the trailing `_` as well.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -535,334 +681,432 @@ applied to the generated HTML input element.
+  places, only the last group of inputs' values will be saved.
+
+    In this example, only the tertiary colors would be passed, and the
+    primary colors would be overridden:
```
Why is this indented?

admusin replied:
This is indented because I've indented by one extra level in all options enumeration lists, the code examples along with their introductory phrases (be they short, such as: "E.g.", "For example" or longer such as the one you indicated here).

The rationale for this was that I found  that some of the option enumeration lists lists (such as Options for Input, Common Options for Specific Input Types as well as Common Options for Select, Checkbox etc.) tend to occupy quite a bit of space and it becomes rather hard to differentiate between what's description/definition and what's example. I found that when landing in the middle of an options list it seems to be easier to visually scan through the list and "pick up" the definitions while ignoring the example if we're back just for a quick check (not a first read). I've fixed now the indenting to always  be 4 spaces per indentation level.
Of course it might be only my subjective impression, and if you think they should stay at the same level of indentation I'll change back the indent for all code/output examples along with their introductory phrase.
Just let me know.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -535,334 +681,432 @@ applied to the generated HTML input element.
+  Disabling ``'hiddenField'`` on the second input group would
+  prevent this behavior.
+
+    You can set a hidden field to a value other than 0, such as 'N'::
```
Why is this indented?

admusin replied:
Idem as above.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -535,334 +681,432 @@ applied to the generated HTML input element.
+
+Creates a ``checkbox`` form element. The widget template used is::
+
+'checkbox' => '<input type="checkbox" name="{{name}}" value="{{value}}"{{attrs}}>'
```
This needs to be indented.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
-    <option value="M">Male</option>
-    <option value="F">Female</option>
-    </select>
+  <select name="field">
```
This needs to be indented.

admusin replied:
Fixed in current commit. I've also went through the whole file and fixed the indent for a bunch of other code samples that were indented by only 3 or 2 spaces; now they should all be at 4 spaces for first indentation level, as you seem to have implicitly suggested. I've also tried to make sure that all extra indentation levels are 4 spaces each, as well.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+      'Value 1' => 'Label 1',
+      'Value 2' => 'Label 2',
+      'Value 3' => 'Label 3'
+  ]);
```
This needs to be indented.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+.. code-block:: html
 
-  .. code-block:: html
+  <select name="field">
```
This needs to be indented.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+For example::
 
-    echo $this->Form->select('field', [
+  $options = [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
-        <option value="Value 2">Label 2</option>
-        <option value="Value 3">Label 3</option>
-    </select>
+  <select name="field">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+.. code-block:: html
 
-  Output:
+  <select name="field">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+  to ``true``. To disable only certain items, assign to ``'disabled'``
+  an array containing the keys of the items to be disabled.
+
+    E.g. ::
```
This is incorrectly indented.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+
+    .. code-block:: html
+
+      <select name="gender">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -888,171 +1135,256 @@ put custom attributes on your radio buttons you can use an expanded format::
+
+    This option also works when ``'multiple'`` is set to ``'checkbox'``::
+
+      $options = [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1073,36 +1405,52 @@ Creating File Inputs
+So start off with a ``create()`` method such as the following::
 
-Next add either of the two lines to your form view file::
+  echo $this->Form->create($document, ['enctype' => 'multipart/form-data']);
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1073,36 +1405,52 @@ Creating File Inputs
 
-    // OR
-    echo $this->Form->file('submittedfile');
+  echo $this->Form->input('submittedfile', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1073,36 +1405,52 @@ Creating File Inputs
-    $this->request->data['submittedfile'] = [
+    $this->request->data['submittedfile']
+
+    //would contain the following array:
```
Space after the `//`

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-    ]);
+For example ::
+
+  <?= $this->form->dateTime('registered') ?>
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+
+.. code-block:: html
+
+  <select name="registered[year]">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-Which would create the following two selects:
+For example ::
+
+  echo $this->Form->datetime('released', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-        <option value="01">January</option>
-        <!-- .. snipped for brevity .. -->
-    </select>
+  <select name="released[year]" class="year-class">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+
+For example::
+
+  // Assuming current year is 2017; this disables day picker, removes empty
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+
+.. code-block:: html
+
+  <select class= "cool-years" name="registered[year]" title="Registration Year">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-            'class' => 'bar-class',
-        ],
-    ]);
+  echo $this->Form->time('released', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-        <option value="30">30</option>
-        <option value="45">45</option>
-    </select>
+  <select name="released[hour]" class="foo-class">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-        'minYear' => 2000,
-        'maxYear' => date('Y')
-    ]);
+  echo $this->Form->year('purchased', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-    <option value="2001">2001</option>
-    <option value="2000">2000</option>
-    </select>
+  <select name="purchased[year]">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-:doc:`/core-libraries/internationalization-and-localization` features.)::
-
-    echo $this->Form->month('mob', ['monthNames' => false]);
+  <select name="mob[month]">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-    echo $this->Form->day('created');
+For example::
+
+  echo $this->Form->day('created', ['empty' => 'Day']);
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
-    ...
-    <option value="31">31</option>
-    </select>
+  <select name="created[day]">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+
+You can create either 12 or 24 hour pickers using the ``'format'`` option::
+
+  echo $this->Form->hour('created', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+For example, if you wanted 10 minutes increments you would do the following::
+
+  // In your view template file
+  echo $this->Form->minute('arrival', [
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1117,272 +1465,504 @@ way PHP handles data passed via file fields
+
+.. code-block:: html
+
+  <select name="arrival[minute]">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1394,67 +1974,157 @@ Output:
-will be used as a class name::
+When you set ``$options`` to a string it will be used as a class name:
+
+::
```
This should be folded with the line above.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1549,56 +2234,127 @@ Will output:
+
+For example::
+
+  // In Templates/Tickets/index.ctp
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1549,56 +2234,127 @@ Will output:
+
+.. code-block:: html
+
+  <form method="post" accept-charset="utf-8" action="/Rtools/tickets/delete/5">
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1657,26 +2424,41 @@ Adding Additional Template Variables to Templates
-    echo $this->Form->input('password', [
-        'templateVars' => ['help' => 'At least 8 characters long.']
-    ]);
+  // Add a template with the help placeholder.
```
This needs to be indented in a multiple of 4.

admusin replied:
Fixed in current commit.

markstory commented on en/views/helpers/form.rst:
```diff
@@ -1657,26 +2424,41 @@ Adding Additional Template Variables to Templates
+
+.. code-block:: html
+
+  <div class="input password">
```
This needs to be indented in a multiple of 4.

admusin commented:
Thanks @markstory for all the feedback; I know how tedious it can be to review somebody else's work.
In all fairness I checked the Documentation Formatting Guide but there was no mention about exact indenting rules to follow so I assumed that the PSR2 style of indenting is not mandatory for the docs. Should I had known it, I would have fixed all that stuff before committing and saved you all that extra work. If nobody does it before me, I'll try to add this mention to the Formatting Guide when I find a moment.

I've fixed pretty much all that was indicated, in the current commit, save for the 2 questions, about quotes and code extra indentation inside enum lists; where I'll need your feedback, when you can find the time, in order to get it right.

inoas commented:
Will that easily merge with https://github.com/cakephp/docs/pull/4559/files ?

antograssiot commented:
@admusing you'll have to solve the conflicts for this to be mergeable.

antograssiot commented:
I'Lloyd try to review it tonight or tomorrow also...

admusin commented:
I've solved the conflicts and committed the solved version.
However, I'm still waiting for Mark's feedback on 2 of the points he raised and where I had asked some additional questions, when he'll find a moment, in order to finish the file.
Since we didn't clear those up yet, I'm not sure that it's ready to be merged yet, even if it doesn't show any conflicts now.

markstory commented:
@admusin I can't find the 2 questions you asked. Where are they?

admusin commented:
Hi @markstory ,

They are in the answers I wrote under your comments. Since now they display as "outdated", I suppose they're not popping up so much anymore and are easy to miss.

First:

> `+By using the ``'type'`` option you can change the HTTP method a form will use::`
> 
>  @markstory 5 days ago Member 
> We don't generally include single quotes around key name as the inline code indicates that it is a key name.
> 
> @admusin 4 days ago • edited Contributor
> 
> I've removed the quotes around this word as indicated.
> I think I've added quotes around pretty much around every key name; if I understand correctly I should remove them everywhere, right?
> As a more particular case, should I remove them even in the enumeration lists (i.e. where we show a list of all the available key names, with their description; I'm asking that since in those places I guess they are meant to indicate what we should literally input in our code and might be helpful to leave no room for guessing for beginners.
> 
> My rationale for adding them was that some trivial key names such "type", "empty", "default", "value" etc. coincide with standard vocabulary words as well as with certain HTML attributes and I thought it would help disambiguate what we refer to, since in certain phrase constructions it can become fuzzy to tell what the author's intended meaning is (it might be more obvious for a veteran Cake user, but then again, they're not the main target audience of the manual as they need it less ).
> Of course, if you think they need to go away, everywhere, I'll remove them all over the page where code examples aren't involved. Just making sure what I should do, for now.

Second:

> ```
> +  places, only the last group of inputs' values will be saved.
> +
> +    In this example, only the tertiary colors would be passed, and the
> +    primary colors would be overridden:
> ```
> @markstory 5 days ago Member
> 
> Why is this indented?
> @admusin 4 days ago Contributor
> 
> This is indented because I've indented by one extra level in all options enumeration lists, the code examples along with their introductory phrases (be they short, such as: "E.g.", "For example" or longer such as the one you indicated here - *In this example, only the tertiary...*).
> 
> The rationale for this was that I found that some of the option enumeration lists lists (such as Options for Input, Common Options for Specific Input Types as well as Common Options for Select, Checkbox etc.) tend to occupy quite a bit of space and it becomes rather hard to differentiate between what's description/definition and what's example. I found that when landing in the middle of an options list it seems to be easier to visually scan through the list and "pick up" the definitions while ignoring the example if we're back just for a quick check (not a first read). I've fixed now the indenting to always be 4 spaces per indentation level.
> Of course it might be only my subjective impression, and if you think they should stay at the same level of indentation I'll change back the indent for all code/output examples along with their introductory phrase.
> Just let me know.

antograssiot commented:
For point 1 you should remove all quotes where the value is already surrounded by backtick

antograssiot commented on en/views/helpers/form.rst:
```diff
@@ -1411,51 +1989,140 @@ Output:
+    If you use a certain model field to generate multiple form fields via
+    ``input()``, and you want the same validation error message displayed for
+    each one, you will probably be better off defining a custom error message
+    inside the respective :ref:`validator rules<create-validators>`.
```
this label doesn't seems to exist and causes the buld to break

admusin replied:
@antograssiot , thanks for the feedback.
Yes, I've just realized that it's mistyped; the real label is "creating-validators".
I will fix it.

So, for the point above (about the quotes) what should I do in the end? I'm a bit fuzzy now, after your edit of your initial comment. :)

Indeed, one of the reasons why I went ahead and added them where they were missing, was the fact that they were already used in the document, on similar words (attribute names, element names, key names etc.) but not in a consistent fashion. So, I was trying to at least make the whole page look the same, not have the same word some times in quotes and some others not.
One other reason is that in most browsers, the simple back ticks yield a pretty mild result (they only change the font) but, visually, the results is not very easy to pick up so they might be missed. On top of that, if you have a word that is surrounded by quotes but not backticks, in the final result your straight quotes will be converted to curly quotes which kind of "alters" a bit their meaning; we tend to assume straight quotes exemplify a value as it should appear in code (e.g. a key name) whereas curly quotes might just mean you want to emphasize or cite something. But it might be just me.

Anyways, in short, if you guys think they should really go away, I can parse the whole page and eliminate them (save for code samples, of course), or else let me know.

antograssiot commented on en/views/helpers/form.rst:
```diff
@@ -16,39 +16,53 @@ Starting a Form
+``$model``:
+
+* An ``Entity`` instance or an iterator will map to
+  `EntityContext <https://api.cakephp.org/3.3/class-Cake.View.Form.EntityContext.html>`_;
```
including those kind of links like that is a bad idea and will get outdated really quickly

admusin replied:
We already have, since quite a while, on this page, the link to the API, inside the List Of Templates section. To be frank I found and still find it highly valuable, as I recently used it quite often. It's true it still points to the 3.2 version of the documentation but that part didn't seem to change in 3.3.

The thing is, that the enumeration list in the beginning of the page, as it is now, is of too little use to any first-comer without any pointer towards those classes; you won't know neither where these classes are. In fact it's not obvious at first view even that they are classes for that matter; when I first arrived on the page I couldn't make heads or tails of this list and figured out that it's referring internal classes, from the small hint in the text of the 3rd option which is *null*, and then it took me a bit of search to actually find them.
The list tells us even less what kind of options/arguments we could use to customize them.

The whole point of the links is to help other users avoid this hurdle and land on the corresponding API page then figure it out, themselves, from there.

I understand the argument about the API code maybe being changed in the future and the link becoming a bit stale; but I would argue that a user savvy enough to start creating custom contexts, will know to search for the last version of the API, starting from the corresponding page where s/he lands.
On the other hand if the API changes that much that these classes are completely removed or change names this section of the page will be probably rewritten anyways.

If you're adamant about removing them, I will though. Don't we have any other means of linking towards an API chapter where the link will seamlessly "migrate" along with the new API versions (let's say, as long as the "target section" name doesn't change, even though its absolute URL might) ?

markstory replied:
If we're linking to the API docs, do we need to duplicate all the options here? Seems redundant and makes the documentation harder to maintain, as we'll need to update both the links and options as they change.

admusin replied:
@markstory 
I've only added the options pertaining to Form Helper.
I didn't add any of the options that can be fed into the entity classes (that's why I added the links to their API pages) as I was thinking it may be too much. Just wanted to provided some doorway for anybody interested to know more about the entities, hence the added links.

Do you suggest that we should remove the links and enumerate the options of each entity instead? Or just simply drop the links?

As I said, I'm not against modifying the file and in the end I'll do what you guys request to make it fit neatly with the rest of the docs.

markstory replied:
Given that the contexts don't have an book docs (as its not something the lay person has to interact with direclty) perhaps an API link is reasonable.

antograssiot commented:
for point 2, you should remove this extra indentation, here is the result i case you missed it:
<img width="851" alt="capture d ecran 2017-01-17 a 13 41 40" src="https://cloud.githubusercontent.com/assets/4977112/22020872/c16579d0-dcba-11e6-873c-5c4b8801b61e.png">

 instead of the actual 

<img width="867" alt="capture d ecran 2017-01-17 a 13 43 11" src="https://cloud.githubusercontent.com/assets/4977112/22020912/ea77bf04-dcba-11e6-98c9-f3c6d24ee55f.png">

antograssiot commented on en/views/helpers/form.rst:
```diff
@@ -91,13 +107,57 @@ Output:
+  provided then it will be autodetected based on the form context.
+  Valid values:
+
+  #. ``'get'`` - Will set the form method to HTTP GET.
```
not sure that we need an ordered list here

admusin replied:
You're right, I'll change that.

antograssiot commented on en/views/helpers/form.rst:
```diff
@@ -1004,85 +1273,158 @@ Output:
+          echo $this->Form->select('gender', $options, ['empty' => true]);
 
-  .. code-block:: html
+      Will output:
```
all those extra indentation needs to be removed
<img width="882" alt="capture d ecran 2017-01-17 a 13 50 15" src="https://cloud.githubusercontent.com/assets/4977112/22021189/1a2d259e-dcbc-11e6-91a1-26340e19f596.png">

admusin replied:
Will do.

admusin commented:
And yes, Mark already mentioned to me about submitting smaller chunks of changes in pull requests as they're easier to review and merge and I agree.

I can see better now, how this may trigger too many back and forth comments for a single request.

admusin commented:
I've fixed the broken link towards Validation page and replaced the ordered sub-list in Options for Form Creation with an unordered one; we should be able to see if at least the auto-checks pass now.
I'll tackle the rest a bit later.

admusin commented:
Removed extra indentation for options' example code blocks.

Still waiting for your final thoughts on the quotes and on the 3 links on the beginning of the page.

dereuromark commented:
@admusin In general, try to make smaller chunks for PRs in the future. This will help to merge things faster.

So are we OK with those improvements as they stand currently?

admusin commented:
@markstory
I've fixed all the requested things so far, except one on which I'll need your input.

We're talking about the quotes around some words.

To summarize:
1. On the original page (leaving code blocks aside) quotes are used together with backquotes on certain keywords that signify options names in Cake, but not all of them. 
2. On some other words, that also signify Cake options, quotes are used without backqoutes, which actually transform them to curly quotes in the final form of the page. 
These make the intended syntax, in the actual code, a bit random and confusing for a newbie (unless they check examples).

What I had done in the beginning was:
1. Adding quotes surrounded by backticks on all Cake option names that occur in text;
2. Adding just backticks on all names of: methods, HTML elements, HTML attributes;
3. Leave simple quotes (that will stay as curly quotes in the final doc) where it made sense that the intended sense wasn't necessarily a property, key, method or attribute name.

*Rationale*
Quite a few of these same words (e.g. "type", "value", "default", etc.) are used in the text in various places as meaning different things:

1. Used as simple English words flowing naturally in the context of the phrase to convey meaning;
2. Used to mean HTML element or attribute names, mentioned when it makes sense in the phrase context;
3. Used to mean Cake keywords signifying method options' names;

_In summary, hoping to clarify things I've used:_

1. For simple English word cases --> no additional quoting or backticks (except in a few cases when I may have leaved the original quotes which are curly quotes on the web page, as it rather emphasized the meaning of the word).
2. For HTML elements and attributes mentioned in text --> just backquotes.
3. For Cake options mentioned in text --> quotes surrounded by backquotes (so that quotes don't become curly and the result looks exactly as in code).

Now, depending on what your own thoughts and preferences are (even if they're completely different from the above), I'll modify the page accordingly; just let me know how you'd like to format each of the 3 cases above.
Thanks.

admusin commented:
Hi @dereuromark, sorry I think I was writing my post above just in the meantime with yours.
Your input of my questions above would be just as welcome.

Once I know what you guys thinks is the better way to go about this, I'll make sure it gets that way and then it will be ready, from my point of view.
Thx.

dereuromark commented:
Is there anything missing or good to go? //CC @markstory
